flowmachine/flowmachine/features/utilities/histogram_aggregation.py

```diff
@@ -0,0 +1,156 @@
+from functools import singledispatch
+
+from ...core.query import Query
+from itertools import count
+from typing import List, Tuple, Optional, Union
+
+
+@singledispatch
+def _get_bins_clause(bins: int) -> Tuple[int, str]:
+    return (
+        bins,
+        f"""
+        SELECT v, 
+            (SELECT lower FROM bounds)+((v-1)*(SELECT (upper-lower)/({bins}) FROM bounds)) as lower, 
+            (SELECT lower FROM bounds)+((v)*(SELECT (upper-lower)/({bins}) FROM bounds)) as upper
+                        FROM generate_series(
+                            1, {bins}) as v
+        """,
+    )
+
+
+@_get_bins_clause.register
+def _(bins: list) -> Tuple[int, str]:
+    return (
+        len(bins),
+        f"""
+                SELECT * FROM (VALUES {",".join(f"({i}, {low}, {high})" for i, low, high in zip(count(), bins[:-1], bins[1:]))}) as t(v, lower, upper)
+                """,
+    )
+
+
+@singledispatch
+def _get_bounds_clause(bounds, value_column: str, metric: Query) -> str:
+    return f"""
+        SELECT max({value_column})::numeric as upper, 
+                min({value_column})::numeric as lower 
+            FROM ({metric.get_query()}) AS to_agg
+        """
+
+
+@_get_bounds_clause.register
+def _(bounds: tuple, value_column: str, metric: Query) -> str:
+    return f"""
+            SELECT {max(bounds)}::numeric as upper, 
+                   {min(bounds)}::numeric  as lower 
+             """
+
+
+class HistogramAggregation(Query):
+    """
+    Compute the histogram of another query.
+
+    Parameters
+    ----------
+    metric : Query
+        Query to build histogram over
+    bins : int, or list of float
+        Either an integer number of equally spaced bins, or a list of bin edges
+    range : tuple of float, default None
+        Optionally supply inclusive lower and upper bounds to build the histogram over. By default, the
+        histogram will cover the whole range of the data.
+    censor : bool, default True
+        Set to False to return results where there are bins with counts below 15
+
+    Notes
+    -----
+    By default, if the count of values for any bin is _below_ 15, then no histogram will be
+    returned.
+
+    Examples
+    --------
+    >>>from flowmachine.features import RadiusOfGyration
+    >>>from flowmachine.features.utilities.histogram_aggregation import HistogramAggregation
+    >>>radius_of_gyration = RadiusOfGyration("2016-01-01", "2016-01-02")
+    >>>histogram = HistogramAggregation(metric=radius_of_gyration, bins=5, censor=False)
+    >>>histogram.head()
+           value  lower_edge  upper_edge
+    0     61    0.000000   70.837717
+    1    123   70.837717  141.675435
+    2    192  141.675435  212.513152
+    3    108  212.513152  283.350869
+    4     15  283.350869  354.188587
+    """
+
+    def __init__(
+        self,
+        *,
+        metric: "Query",
+        bins: Union[List[float], int],
+        range: Optional[Tuple[float, float]] = None,
```

Comment:
Probably not a good idea to use a keyword as an argument name

Reply:
Or rather, a Python built-in function name

Reply:
Yeah, I agree - but following the numpy function signature: https://docs.scipy.org/doc/numpy/reference/generated/numpy.histogram.html

Reply:
Ah, I see. Fair enough.